mmdet/core/bbox/assigners/center_region_assigner.py

```diff
@@ -0,0 +1,301 @@
+            #    have high priority to assign the pixel.
+            gt_areas = bboxes_area(gt_bboxes)
+            # Rank all gt bbox areas. Smaller objects has larger priority
+            _, sort_idx = gt_areas.sort(descending=True)
+            assigned_gt_ids, pixels_in_gt_shadow = \
+                self.assign_one_hot_gt_indices(is_bbox_in_gt_core,
+                                               is_bbox_in_gt_shadow,
+                                               gt_priority=sort_idx)
+
+        if gt_bboxes_ignore is not None and gt_bboxes_ignore.numel() > 0:
+            # No ground truth or boxes, return empty assignment
+            gt_bboxes_ignore = scale_boxes(
+                gt_bboxes_ignore, scale=self.ignore_gt_scale)
+            is_bbox_in_ignored_gts = is_located_in(bbox_centers,
+                                                   gt_bboxes_ignore)
+            is_bbox_in_ignored_gts = is_bbox_in_ignored_gts.any(dim=1)
+            assigned_gt_ids[is_bbox_in_ignored_gts] = -1
+
+        # 4. Assign prior bboxes with class label according to its gt id.
+        assigned_labels = None
+        shadowed_pixel_labels = None
+        if gt_labels is not None:
+            assigned_labels = assigned_gt_ids.new_full((num_bboxes, ), -1)
+            pos_inds = torch.nonzero(assigned_gt_ids > 0).squeeze()
+            if pos_inds.numel() > 0:
+                assigned_labels[pos_inds] = gt_labels[assigned_gt_ids[pos_inds]
+                                                      - 1]
+            # 5. Find pixels lying in the shadow of an object
+            shadowed_pixel_labels = pixels_in_gt_shadow.clone()
+            if pixels_in_gt_shadow.numel() > 0:
+                pixel_idx, gt_idx =\
+                    pixels_in_gt_shadow[:, 0], pixels_in_gt_shadow[:, 1]
+                assert (assigned_gt_ids[pixel_idx] != gt_idx).all(), \
+                    'Some pixels are dually assigned to ignore and gt!'
+                shadowed_pixel_labels[:, 1] = gt_labels[gt_idx - 1]
+                # When a pixel is both positive and shadowed, set it as shadow.
+                override = (
+                    assigned_labels[pixel_idx] == shadowed_pixel_labels[:, 1])
+                assigned_labels[pixel_idx[override]] = -1
+                assigned_gt_ids[pixel_idx[override]] = 0
+
+        assign_result = AssignResult(
+            num_gts, assigned_gt_ids, None, labels=assigned_labels)
+        # Add shadowed_labels as assign_result property. Shape: (num_shadow, 2)
+        assign_result.set_extra_property('shadowed_labels',
+                                         shadowed_pixel_labels)
+        return assign_result
+
+    def assign_one_hot_gt_indices(self,
+                                  is_bbox_in_gt_core,
+                                  is_bbox_in_gt_shadow,
+                                  gt_priority=None):
+        """Assign only one gt index to each prior box
+
+        Smaller gts have higher priority
```

Comment:
Is this a common practice? Or just in FSAF?

Reply:
It is common practice to assign only one gt index to each prior bbox, while it is mainly FSAF that assigns the anchor with gt of smaller area.

Reply:
If only FSAF prefers small gts, we may set this step as optional, or just name the assigner FSAFAssigner to differ it from other center region based assigners.

Reply:
This assigner is actually designed to be of general use. The given gt_priority can be specified by the user to treat overlapping gts. Maybe I should modify the docstring instead.

Reply:
Better to make the preference of small gts optional, which can be more general.

Reply:
The gt_priority is now obtained by a function, which can be overloaded by the user.